litr/src/main/java/com/linkedin/android/litr/render/PassthroughSoftwareRenderer.java

```diff
@@ -49,33 +50,56 @@ public void renderFrame(@Nullable Frame frame, long presentationTimeNs) {
             Log.e(TAG, "Null or empty input frame provided");
             return;
         }
-        int tag = encoder.dequeueInputFrame(FRAME_WAIT_TIMEOUT);
-        if (tag >= 0) {
-            Frame outputFrame = encoder.getInputFrame(tag);
-            if (outputFrame == null) {
-                Log.e(TAG, "No input frame returned by an encoder, dropping a frame");
-                return;
-            }
 
-            ByteBuffer inputBuffer = frame.buffer.asReadOnlyBuffer();
-            inputBuffer.rewind();
-            outputFrame.buffer.put(inputBuffer);
+        ArrayList<Frame> outputFrames = new ArrayList<>();
+        ByteBuffer inputBuffer = null;
+        boolean isFirst = true;
+        boolean isContinue;
+        do {
+            isContinue = false;
+            int tag = encoder.dequeueInputFrame(FRAME_WAIT_TIMEOUT);
+            if (tag >= 0) {
+                Frame outputFrame = encoder.getInputFrame(tag);
+                if (outputFrame == null) {
+                    Log.e(TAG, "No input frame returned by an encoder, dropping a frame");
+                    return;
+                }
+
+                if (isFirst) {
+                    isFirst = false;
+                    inputBuffer = frame.buffer.asReadOnlyBuffer();
+                    inputBuffer.rewind();
+                }
+
+                int capacity = Math.min(outputFrame.buffer.remaining(), inputBuffer.remaining());
+                byte[] inputBytes = new byte[capacity];
+                inputBuffer.get(inputBytes, 0, capacity);
+
+                outputFrame.buffer.put(inputBytes);
+
+                outputFrame.bufferInfo.set(
+                        0,
+                        capacity,
+                        TimeUnit.NANOSECONDS.toMicros(presentationTimeNs),
+                        frame.bufferInfo.flags);
+                outputFrames.add(outputFrame);
+
+                isContinue = inputBuffer.hasRemaining();
+            } else {
+                switch (tag) {
+                    case MediaCodec.INFO_TRY_AGAIN_LATER:
+                        Log.e(TAG, "Encoder input frame timeout, dropping a frame");
+                        break;
+                    default:
+                        Log.e(TAG, "Unhandled value " + tag + " when receiving decoded input frame");
+                        break;
+                }
+            }
+        }
+        while (isContinue);
 
-            outputFrame.bufferInfo.set(
-                    0,
-                    frame.bufferInfo.size,
-                    TimeUnit.NANOSECONDS.toMicros(presentationTimeNs),
-                    frame.bufferInfo.flags);
+        for (Frame outputFrame : outputFrames) {
```

Comment:
Why do we need a separate loop for this? Can't we do this in the `do while` loop and enqueue encoder input frame as soon as we fill it with data?